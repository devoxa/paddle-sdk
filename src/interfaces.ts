export type PaddleSdkSubscriptionStatus =
  | 'ACTIVE'
  | 'TRIALING'
  | 'PAST_DUE'
  | 'PAUSED'
  | 'CANCELLED'

export type PaddleSdkCurrency =
  | 'USD'
  | 'EUR'
  | 'GBP'
  | 'ARS'
  | 'AUD'
  | 'BRL'
  | 'CAD'
  | 'CHF'
  | 'CNY'
  | 'CZK'
  | 'DKK'
  | 'HKD'
  | 'HUF'
  | 'INR'
  | 'JPY'
  | 'KRW'
  | 'MXN'
  | 'NOK'
  | 'NZD'
  | 'PLN'
  | 'RUB'
  | 'SEK'
  | 'SGD'
  | 'THB'
  | 'TWD'
  | 'ZAR'

export type PaddleSdkCountry =
  | 'AF'
  | 'AL'
  | 'DZ'
  | 'AS'
  | 'AD'
  | 'AO'
  | 'AI'
  | 'AG'
  | 'AR'
  | 'AM'
  | 'AW'
  | 'AU'
  | 'AT'
  | 'AZ'
  | 'BS'
  | 'BH'
  | 'BD'
  | 'BB'
  | 'BY'
  | 'BE'
  | 'BZ'
  | 'BJ'
  | 'BM'
  | 'BT'
  | 'BO'
  | 'BA'
  | 'BW'
  | 'BV'
  | 'BR'
  | 'IO'
  | 'VG'
  | 'BN'
  | 'BG'
  | 'BF'
  | 'BI'
  | 'KH'
  | 'CM'
  | 'CA'
  | 'CV'
  | 'KY'
  | 'CF'
  | 'TD'
  | 'CL'
  | 'CN'
  | 'CX'
  | 'CC'
  | 'CO'
  | 'KM'
  | 'CG'
  | 'CK'
  | 'CR'
  | 'CI'
  | 'HR'
  | 'CU'
  | 'CW'
  | 'CY'
  | 'CZ'
  | 'DK'
  | 'DJ'
  | 'DM'
  | 'DO'
  | 'EC'
  | 'EG'
  | 'SV'
  | 'GQ'
  | 'ER'
  | 'EE'
  | 'ET'
  | 'FK'
  | 'FO'
  | 'FJ'
  | 'FI'
  | 'FR'
  | 'GF'
  | 'PF'
  | 'TF'
  | 'GA'
  | 'GM'
  | 'GE'
  | 'DE'
  | 'GH'
  | 'GI'
  | 'GR'
  | 'GL'
  | 'GD'
  | 'GP'
  | 'GU'
  | 'GT'
  | 'GG'
  | 'GN'
  | 'GW'
  | 'GY'
  | 'HT'
  | 'HM'
  | 'VA'
  | 'HN'
  | 'HK'
  | 'HU'
  | 'IS'
  | 'IN'
  | 'ID'
  | 'IR'
  | 'IQ'
  | 'IE'
  | 'IL'
  | 'IT'
  | 'JM'
  | 'JP'
  | 'JE'
  | 'JO'
  | 'KZ'
  | 'KE'
  | 'KI'
  | 'KW'
  | 'KG'
  | 'LA'
  | 'LV'
  | 'LB'
  | 'LS'
  | 'LR'
  | 'LY'
  | 'LI'
  | 'LT'
  | 'LU'
  | 'MO'
  | 'MK'
  | 'MG'
  | 'MW'
  | 'MY'
  | 'MV'
  | 'ML'
  | 'MT'
  | 'MH'
  | 'MQ'
  | 'MR'
  | 'MU'
  | 'YT'
  | 'MX'
  | 'FM'
  | 'MD'
  | 'MC'
  | 'MN'
  | 'ME'
  | 'MS'
  | 'MA'
  | 'MZ'
  | 'MM'
  | 'NA'
  | 'NR'
  | 'NP'
  | 'NL'
  | 'AN'
  | 'NC'
  | 'NZ'
  | 'NI'
  | 'NE'
  | 'NG'
  | 'NU'
  | 'NF'
  | 'KP'
  | 'MP'
  | 'NO'
  | 'OM'
  | 'PK'
  | 'PW'
  | 'PS'
  | 'PA'
  | 'PG'
  | 'PY'
  | 'PE'
  | 'PH'
  | 'PN'
  | 'PL'
  | 'PT'
  | 'PR'
  | 'QA'
  | 'RS'
  | 'RE'
  | 'RO'
  | 'RU'
  | 'RW'
  | 'GS'
  | 'SH'
  | 'KN'
  | 'LC'
  | 'PM'
  | 'VC'
  | 'WS'
  | 'SM'
  | 'ST'
  | 'SA'
  | 'SN'
  | 'SC'
  | 'SL'
  | 'SG'
  | 'SK'
  | 'SI'
  | 'SB'
  | 'SO'
  | 'ZA'
  | 'KR'
  | 'ES'
  | 'LK'
  | 'SD'
  | 'SR'
  | 'SJ'
  | 'SZ'
  | 'SE'
  | 'CH'
  | 'SY'
  | 'TW'
  | 'TJ'
  | 'TZ'
  | 'TH'
  | 'TL'
  | 'TG'
  | 'TK'
  | 'TO'
  | 'TT'
  | 'TN'
  | 'TR'
  | 'TM'
  | 'TC'
  | 'TV'
  | 'VI'
  | 'UG'
  | 'UA'
  | 'AE'
  | 'GB'
  | 'US'
  | 'UM'
  | 'UY'
  | 'UZ'
  | 'VU'
  | 'VE'
  | 'VN'
  | 'WF'
  | 'EH'
  | 'YE'
  | 'ZM'
  | 'ZW'
  | 'AU'
  | 'CA'
  | 'FR'
  | 'DE'
  | 'IN'
  | 'IT'
  | 'NL'
  | 'ES'
  | 'GB'
  | 'US'

export type PaddleSdkPausedReason = 'DELINQUENT' | 'VOLUNTARY'

export type PaddleSdkPaymentMethod = 'CARD' | 'PAYPAL' | 'APPLE_PAY' | 'WIRE_TRANSFER' | 'FREE'

export type PaddleSdkSubscriptionCreatedAlert<T> = {
  alert_id: number
  alert_name: 'SUBSCRIPTION_CREATED'
  cancel_url: string
  checkout_id: string
  currency: PaddleSdkCurrency
  email: string
  event_time: Date
  marketing_consent: boolean
  next_bill_date: Date
  passthrough: T
  quantity: number
  source: string
  status: PaddleSdkSubscriptionStatus
  subscription_id: number
  subscription_plan_id: number
  unit_price: number
  update_url: string
  user_id: number
}

export type PaddleSdkSubscriptionUpdatedAlert<T> = {
  alert_id: number
  alert_name: 'SUBSCRIPTION_UPDATED'
  cancel_url: string
  checkout_id: string
  currency: PaddleSdkCurrency
  email: string
  event_time: Date
  marketing_consent: boolean
  new_next_bill_date: Date
  new_price: number
  new_quantity: number
  new_status: PaddleSdkSubscriptionStatus
  new_subscription_plan_id: number
  new_unit_price: number
  old_next_bill_date: Date
  old_price: number
  old_quantity: number
  old_status: PaddleSdkSubscriptionStatus
  old_subscription_plan_id: number
  old_unit_price: number
  passthrough: T
  paused_at: Date | null
  paused_from: Date | null
  paused_reason: PaddleSdkPausedReason | null
  subscription_id: number
  update_url: string
  user_id: number
}

export type PaddleSdkSubscriptionCancelledAlert<T> = {
  alert_id: number
  alert_name: 'SUBSCRIPTION_CANCELLED'
  cancellation_effective_date: Date
  checkout_id: string
  currency: PaddleSdkCurrency
  email: string
  event_time: Date
  marketing_consent: boolean
  passthrough: T
  quantity: number
  status: PaddleSdkSubscriptionStatus
  subscription_id: number
  subscription_plan_id: number
  unit_price: number
  user_id: number
}

export type PaddleSdkSubscriptionPaymentSucceededAlert<T> = {
  alert_id: number
  alert_name: 'SUBSCRIPTION_PAYMENT_SUCCEEDED'
  balance_currency: PaddleSdkCurrency
  balance_earnings: number
  balance_fee: number
  balance_gross: number
  balance_tax: number
  checkout_id: string
  country: string
  coupon: string
  currency: PaddleSdkCurrency
  customer_name: string
  earnings: number
  email: string
  event_time: Date
  fee: number
  initial_payment: boolean
  instalments: number
  marketing_consent: boolean
  next_bill_date: Date
  next_payment_amount: number
  order_id: string
  passthrough: T
  payment_method: PaddleSdkPaymentMethod
  payment_tax: number
  plan_name: string
  quantity: number
  receipt_url: string
  sale_gross: number
  status: PaddleSdkSubscriptionStatus
  subscription_id: number
  subscription_payment_id: number
  subscription_plan_id: number
  unit_price: number
  user_id: number
}

export type PaddleSdkCreateProductPayLinkRequest<T> = {
  product_id?: number
  title?: string
  webhook_url?: string
  prices?: Array<string>
  recurring_prices?: Array<string>
  trial_days?: number
  custom_message?: string
  coupon_code?: string
  discountable?: 0 | 1
  image_url?: string
  return_url?: string
  quantity_variable?: 0 | 1
  quantity?: number
  expires?: string
  affiliates?: Array<string>
  recurring_affiliate_limit?: number
  marketing_consent?: 0 | 1
  customer_email?: string
  customer_country?: string
  customer_postcode?: string
  passthrough?: T
  vat_number?: string
  vat_company_name?: string
  vat_street?: string
  vat_city?: string
  vat_state?: string
  vat_country?: string
  vat_postcode?: string
}

export type PaddleSdkCreateProductPayLinkResponse = {
  url: string
}

export type PaddleSdkListSubscriptionsRequest = {
  subscription_id?: string
  plan_id?: string
  state?: 'active' | 'past_due' | 'trialing' | 'paused' | 'deleted'
  page?: number
  results_per_page?: number
}

export type PaddleSdkListSubscriptionsResponse = Array<{
  subscription_id: number
  plan_id: number
  user_id: number
  user_email: string
  marketing_consent: boolean
  status: PaddleSdkSubscriptionStatus
  signup_date: Date
  update_url: string
  cancel_url: string
  paused_at: Date
  paused_from: Date
  payment_information:
    | {
        payment_method: 'card'
        card_type: string
        last_four_digits: string
        expiry_date: Date
      }
    | {
        payment_method: 'paypal'
      }
  last_payment: {
    amount: number
    currency: PaddleSdkCurrency
    date: Date
  }
  next_payment: {
    amount: number
    currency: PaddleSdkCurrency
    date: Date
  } | null
}>

export type PaddleSdkUpdateSubscriptionRequest<T> = {
  subscription_id: number
  quantity?: number
  currency?: string
  recurring_price?: number
  bill_immediately?: boolean
  plan_id?: number
  prorate?: boolean
  keep_modifiers?: boolean
  passthrough?: T
  pause?: boolean
}

export type PaddleSdkUpdateSubscriptionResponse = {
  subscription_id: number
  user_id: number
  plan_id: number
  next_payment: {
    amount: number
    currency: string
    date: string
  }
}

export type PaddleSdkCancelSubscriptionRequest = {
  subscription_id: number
}

export type PaddleSdkCancelSubscriptionResponse = void

export type PaddleSdkCreateSubscriptionModifierRequest = {
  subscription_id: number
  modifier_recurring?: true | false
  modifier_amount: number
  modifier_description?: string
}

export type PaddleSdkCreateSubscriptionModifierResponse = {
  subscription_id: number
  modifier_id: number
}
