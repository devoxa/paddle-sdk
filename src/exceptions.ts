export class PaddleSdkException extends Error {}
