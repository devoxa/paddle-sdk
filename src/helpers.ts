import { serialize as phpSerialize } from 'php-serialize'
import dayjs from 'dayjs'
import customParseFormat from 'dayjs/plugin/customParseFormat'
import utc from 'dayjs/plugin/utc'
import FormData from 'form-data'

dayjs.extend(customParseFormat)
dayjs.extend(utc)

export function stableSerialize(object: { [key: string]: any }): string {
  // 1) Sort the object alphabetically by it's keys
  object = sortByKey(object)

  // 2) Encode arrays in their string form: `[1, 2, 3]` -> `'1, 2, 3'`
  // 3) Encode any non-strings as their JSON stringified version: `3` -> `'3'`
  for (const property in object) {
    if (object.hasOwnProperty(property) && typeof object[property] !== 'string') {
      if (Array.isArray(object[property])) {
        object[property] = object[property].join(', ')
      } else {
        object[property] = JSON.stringify(object[property])
      }
    }
  }

  // 4) Serialize in the way that PHP would
  return phpSerialize(object)
}

function sortByKey(object: { [key: string]: any }) {
  const keys = Object.keys(object).sort()

  const sortedObject: { [key: string]: any } = {}
  for (const i in keys) {
    sortedObject[keys[i]] = object[keys[i]]
  }

  return sortedObject
}

export function parseUtcDate(dateString: string, type: 'DATE' | 'DATE_TIME' | 'EXPIRY_DATE'): Date {
  switch (type) {
    case 'DATE':
      return dayjs.utc(dateString, 'YYYY-MM-DD').toDate()
    case 'DATE_TIME':
      return dayjs.utc(dateString, 'YYYY-MM-DD HH:mm:ss').toDate()
    case 'EXPIRY_DATE':
      return dayjs.utc(dateString, 'MM/YYYY').toDate()
  }
}

export function objectToFormData(object: { [key: string]: any }): FormData {
  const formData = new FormData()

  Object.entries(object).map(([key, value]) => {
    formData.append(key, value.toString())
  })

  return formData
}
