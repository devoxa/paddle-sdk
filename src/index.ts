import { createVerify } from 'crypto'
import fetch from 'node-fetch'
import { stableSerialize, parseUtcDate, objectToFormData } from './helpers'
import {
  RawPaddleWebhookAlert,
  RawPaddleSubscriptionCreatedAlert,
  RawPaddleSubscriptionUpdatedAlert,
  RawPaddleSubscriptionCancelledAlert,
  RawPaddleSubscriptionPaymentSucceededAlert,
} from './__generated__/webhook-alert-interfaces'
import { PaddleSdkException, PaddleSdkApiException } from './exceptions'
import {
  PaddleSdkSubscriptionCreatedAlert,
  PaddleSdkSubscriptionUpdatedAlert,
  PaddleSdkSubscriptionCancelledAlert,
  PaddleSdkSubscriptionPaymentSucceededAlert,
  PaddleSdkCreateProductPayLinkRequest,
  PaddleSdkListSubscriptionsRequest,
  PaddleSdkUpdateSubscriptionRequest,
  PaddleSdkCancelSubscriptionRequest,
  PaddleSdkCreateSubscriptionModifierRequest,
  PaddleSdkCreateProductPayLinkResponse,
  PaddleSdkListSubscriptionsResponse,
  PaddleSdkUpdateSubscriptionResponse,
  PaddleSdkCancelSubscriptionResponse,
  PaddleSdkCreateSubscriptionModifierResponse,
  PaddleSdkSubscriptionStatus,
  PaddleSdkCurrency,
  PaddleSdkPausedReason,
  PaddleSdkPaymentMethod,
} from './interfaces'
import {
  RawPaddlePostProductGeneratePayLinkRequest,
  RawPaddlePostProductGeneratePayLinkResponse,
  PADDLE_PRODUCT_GENERATE_PAY_LINK,
  RawPaddlePostSubscriptionUsersUpdateRequest,
  RawPaddlePostSubscriptionUsersUpdateResponse,
  PADDLE_SUBSCRIPTION_USERS_UPDATE,
  RawPaddlePostSubscriptionModifiersCreateRequest,
  RawPaddlePostSubscriptionModifiersCreateResponse,
  PADDLE_SUBSCRIPTION_MODIFIERS_CREATE,
  RawPaddlePostSubscriptionUsersCancelRequest,
  RawPaddlePostSubscriptionUsersCancelResponse,
  PADDLE_SUBSCRIPTION_USERS_CANCEL,
  RawPaddlePostSubscriptionUsersRequest,
  RawPaddlePostSubscriptionUsersResponse,
  PADDLE_SUBSCRIPTION_USERS,
} from './__generated__/api-route-interfaces'
import { encrypt, decrypt } from '@devoxa/aes-encryption'

export * from './interfaces'

export interface PaddleSdkOptions {
  publicKey: string
  vendorId: number
  vendorAuthCode: string
  passthroughEncryptionKey: string
}

export class PaddleSdk<TPassthrough = any> {
  private readonly publicKey: string
  private readonly vendorId: number
  private readonly vendorAuthCode: string
  private readonly passthroughEncryptionKey: string

  constructor(options: PaddleSdkOptions) {
    if (!options.publicKey) {
      throw new PaddleSdkException('PaddleSdk was called without a publicKey')
    }

    if (!options.vendorId) {
      throw new PaddleSdkException('PaddleSdk was called without a vendorId')
    }

    if (!options.vendorAuthCode) {
      throw new PaddleSdkException('PaddleSdk was called without a vendorAuthCode')
    }

    if (!options.passthroughEncryptionKey || options.passthroughEncryptionKey.length !== 32) {
      throw new PaddleSdkException('PaddleSdk was called with an invalid passthroughEncryptionKey')
    }

    this.publicKey = options.publicKey
    this.vendorId = options.vendorId
    this.vendorAuthCode = options.vendorAuthCode
    this.passthroughEncryptionKey = options.passthroughEncryptionKey
  }

  verifyWebhookAlert(body: any): body is RawPaddleWebhookAlert {
    if (typeof body !== 'object') return false

    const { p_signature: signature, ...postBodyRest } = body || {}
    if (!signature || typeof signature !== 'string') return false

    const serializedPostBody = stableSerialize(postBodyRest)

    const verifier = createVerify('sha1')
    verifier.update(serializedPostBody)
    verifier.end()

    return verifier.verify(this.publicKey, signature, 'base64')
  }

  parseWebhookAlert(body: any) {
    if (!this.verifyWebhookAlert(body)) {
      throw new PaddleSdkException('Failed validating alert body')
    }

    switch (body.alert_name) {
      case 'subscription_created':
        return this.parseSubscriptionCreatedWebhookAlert(body)
      case 'subscription_updated':
        return this.parseSubscriptionUpdatedWebhookAlert(body)
      case 'subscription_cancelled':
        return this.parseSubscriptionCancelledWebhookAlert(body)
      case 'subscription_payment_succeeded':
        return this.parseSubscriptionPaymentSucceededWebhookAlert(body)
    }

    throw new PaddleSdkException(
      `Implementation missing: Can not parse alert of type ${body.alert_name}`
    )
  }

  private stringifyPassthrough(passthrough: any) {
    return encrypt(this.passthroughEncryptionKey, JSON.stringify(passthrough))
  }

  private parsePassthrough(passthrough: string): TPassthrough {
    try {
      return JSON.parse(decrypt(this.passthroughEncryptionKey, passthrough))
    } catch (err) {
      throw new PaddleSdkException('Failed parsing passthrough: ' + err.message)
    }
  }

  private parseStatus(
    status: RawPaddleSubscriptionCreatedAlert['status']
  ): PaddleSdkSubscriptionStatus {
    switch (status) {
      case 'active':
        return 'ACTIVE'
      case 'trialing':
        return 'TRIALING'
      case 'past_due':
        return 'PAST_DUE'
      case 'paused':
        return 'PAUSED'
      case 'deleted':
        return 'CANCELLED'
    }
  }

  private parsePausedReason(
    pausedReason: RawPaddleSubscriptionUpdatedAlert['paused_reason']
  ): PaddleSdkPausedReason {
    switch (pausedReason) {
      case 'delinquent':
        return 'DELINQUENT'
      case 'voluntary':
        return 'VOLUNTARY'
    }
  }

  private parsePaymentMethod(
    paymentMethod: RawPaddleSubscriptionPaymentSucceededAlert['payment_method']
  ): PaddleSdkPaymentMethod {
    switch (paymentMethod) {
      case 'card':
        return 'CARD'
      case 'paypal':
        return 'PAYPAL'
      case 'apple-pay':
        return 'APPLE_PAY'
      case 'wire-transfer':
        return 'WIRE_TRANSFER'
      case 'free':
        return 'FREE'
    }
  }

  private parseSubscriptionCreatedWebhookAlert(
    body: RawPaddleSubscriptionCreatedAlert
  ): PaddleSdkSubscriptionCreatedAlert<TPassthrough> {
    return {
      alert_id: parseInt(body.alert_id),
      alert_name: 'SUBSCRIPTION_CREATED',
      cancel_url: body.cancel_url,
      checkout_id: body.checkout_id,
      currency: body.currency as PaddleSdkCurrency,
      email: body.email,
      event_time: parseUtcDate(body.event_time, 'DATE_TIME'),
      marketing_consent: body.marketing_consent === '1',
      next_bill_date: parseUtcDate(body.next_bill_date, 'DATE'),
      passthrough: this.parsePassthrough(body.passthrough),
      quantity: parseInt(body.quantity),
      source: body.source,
      status: this.parseStatus(body.status),
      subscription_id: parseInt(body.subscription_id),
      subscription_plan_id: parseInt(body.subscription_plan_id),
      unit_price: parseFloat(body.unit_price),
      update_url: body.update_url,
      user_id: parseInt(body.user_id),
    }
  }

  private parseSubscriptionUpdatedWebhookAlert(
    body: RawPaddleSubscriptionUpdatedAlert
  ): PaddleSdkSubscriptionUpdatedAlert<TPassthrough> {
    return {
      alert_id: parseInt(body.alert_id),
      alert_name: 'SUBSCRIPTION_UPDATED',
      cancel_url: body.cancel_url,
      checkout_id: body.checkout_id,
      currency: body.currency as PaddleSdkCurrency,
      email: body.email,
      event_time: parseUtcDate(body.event_time, 'DATE_TIME'),
      marketing_consent: body.marketing_consent === '1',
      new_next_bill_date: parseUtcDate(body.next_bill_date, 'DATE'),
      new_price: parseFloat(body.new_price),
      new_quantity: parseInt(body.new_quantity),
      new_status: this.parseStatus(body.status),
      new_subscription_plan_id: parseInt(body.subscription_plan_id),
      new_unit_price: parseFloat(body.new_unit_price),
      old_next_bill_date: parseUtcDate(body.old_next_bill_date, 'DATE'),
      old_price: parseFloat(body.old_price),
      old_quantity: parseInt(body.old_quantity),
      old_status: this.parseStatus(body.old_status),
      old_subscription_plan_id: parseInt(body.old_subscription_plan_id),
      old_unit_price: parseFloat(body.old_unit_price),
      passthrough: this.parsePassthrough(body.passthrough),
      paused_at: body.paused_at ? parseUtcDate(body.paused_at, 'DATE_TIME') : null,
      paused_from: body.paused_from ? parseUtcDate(body.paused_from, 'DATE_TIME') : null,
      paused_reason: body.paused_reason ? this.parsePausedReason(body.paused_reason) : null,
      subscription_id: parseInt(body.subscription_id),
      update_url: body.update_url,
      user_id: parseInt(body.user_id),
    }
  }

  private parseSubscriptionCancelledWebhookAlert(
    body: RawPaddleSubscriptionCancelledAlert
  ): PaddleSdkSubscriptionCancelledAlert<TPassthrough> {
    return {
      alert_id: parseInt(body.alert_id),
      alert_name: 'SUBSCRIPTION_CANCELLED',
      cancellation_effective_date: parseUtcDate(body.cancellation_effective_date, 'DATE'),
      checkout_id: body.checkout_id,
      currency: body.currency as PaddleSdkCurrency,
      email: body.email,
      event_time: parseUtcDate(body.event_time, 'DATE_TIME'),
      marketing_consent: body.marketing_consent === '1',
      passthrough: this.parsePassthrough(body.passthrough),
      quantity: parseInt(body.quantity),
      status: this.parseStatus(body.status),
      subscription_id: parseInt(body.subscription_id),
      subscription_plan_id: parseInt(body.subscription_plan_id),
      unit_price: parseFloat(body.unit_price),
      user_id: parseInt(body.user_id),
    }
  }

  private parseSubscriptionPaymentSucceededWebhookAlert(
    body: RawPaddleSubscriptionPaymentSucceededAlert
  ): PaddleSdkSubscriptionPaymentSucceededAlert<TPassthrough> {
    return {
      alert_id: parseInt(body.alert_id),
      alert_name: 'SUBSCRIPTION_PAYMENT_SUCCEEDED',
      balance_currency: body.balance_currency as PaddleSdkCurrency,
      balance_earnings: parseFloat(body.balance_earnings),
      balance_fee: parseFloat(body.balance_fee),
      balance_gross: parseFloat(body.balance_gross),
      balance_tax: parseFloat(body.balance_tax),
      checkout_id: body.checkout_id,
      country: body.country,
      coupon: body.coupon,
      currency: body.currency as PaddleSdkCurrency,
      customer_name: body.customer_name,
      earnings: parseFloat(body.earnings),
      email: body.email,
      event_time: parseUtcDate(body.event_time, 'DATE_TIME'),
      fee: parseFloat(body.fee),
      initial_payment: body.initial_payment === '1',
      instalments: parseInt(body.instalments),
      marketing_consent: body.marketing_consent === '1',
      next_bill_date: parseUtcDate(body.next_bill_date, 'DATE'),
      next_payment_amount: parseFloat(body.next_payment_amount),
      order_id: body.order_id,
      passthrough: this.parsePassthrough(body.passthrough),
      payment_method: this.parsePaymentMethod(body.payment_method),
      payment_tax: parseFloat(body.payment_tax),
      plan_name: body.plan_name,
      quantity: parseInt(body.quantity),
      receipt_url: body.receipt_url,
      sale_gross: parseFloat(body.sale_gross),
      status: this.parseStatus(body.status),
      subscription_id: parseInt(body.subscription_id),
      subscription_payment_id: parseInt(body.subscription_payment_id),
      subscription_plan_id: parseInt(body.subscription_plan_id),
      unit_price: parseFloat(body.unit_price),
      user_id: parseInt(body.user_id),
    }
  }

  private async apiRequest<TRequest, TResponse>(
    url: string,
    method: 'GET' | 'POST',
    body: TRequest
  ): Promise<TResponse> {
    const formData = objectToFormData({
      vendor_id: this.vendorId,
      vendor_auth_code: this.vendorAuthCode,
      ...body,
    })

    const response = await fetch(url, { method, body: formData })
    const json = await response.json()

    // Turn errors from the Paddle API into a unique exception with the error message
    if (json.success === false) {
      throw new PaddleSdkApiException(json.error.message)
    }

    return json.response
  }

  async createProductPayLink(
    data: PaddleSdkCreateProductPayLinkRequest<TPassthrough>
  ): Promise<PaddleSdkCreateProductPayLinkResponse> {
    return this.apiRequest<
      RawPaddlePostProductGeneratePayLinkRequest,
      RawPaddlePostProductGeneratePayLinkResponse
    >(PADDLE_PRODUCT_GENERATE_PAY_LINK.url, PADDLE_PRODUCT_GENERATE_PAY_LINK.method, {
      ...data,
      passthrough: this.stringifyPassthrough(data.passthrough),
    })
  }

  async listSubscriptions(
    data: PaddleSdkListSubscriptionsRequest
  ): Promise<PaddleSdkListSubscriptionsResponse> {
    function formatPaymentInformation(
      paymentInformation: RawPaddlePostSubscriptionUsersResponse[0]['payment_information']
    ) {
      if (paymentInformation.payment_method !== 'card') {
        // Only card & paypal are enabled for subscriptions, but there might be more options in the future
        return { payment_method: paymentInformation.payment_method }
      }

      return {
        payment_method: 'card' as const,
        card_type: paymentInformation.card_type, // TODO Would be nice to have strict types here
        last_four_digits: paymentInformation.last_four_digits,
        expiry_date: parseUtcDate(paymentInformation.expiry_date, 'EXPIRY_DATE'),
      }
    }

    function formatPayment(payment: RawPaddlePostSubscriptionUsersResponse[0]['last_payment']) {
      return {
        amount: payment.amount,
        currency: payment.currency as PaddleSdkCurrency,
        date: parseUtcDate(payment.date, 'DATE'),
      }
    }

    const formatSubscription = (subscription: RawPaddlePostSubscriptionUsersResponse[0]) => {
      return {
        subscription_id: subscription.subscription_id,
        plan_id: subscription.plan_id,
        user_id: subscription.user_id,
        user_email: subscription.user_email,
        marketing_consent: subscription.marketing_consent,
        status: this.parseStatus(subscription.state),
        signup_date: parseUtcDate(subscription.signup_date, 'DATE_TIME'),
        update_url: subscription.update_url,
        cancel_url: subscription.cancel_url,
        paused_at: parseUtcDate(subscription.paused_at, 'DATE_TIME'),
        paused_from: parseUtcDate(subscription.paused_from, 'DATE_TIME'),
        payment_information: formatPaymentInformation(subscription.payment_information),
        last_payment: formatPayment(subscription.last_payment),
        next_payment: subscription.next_payment ? formatPayment(subscription.next_payment) : null,
      }
    }

    return this.apiRequest<
      RawPaddlePostSubscriptionUsersRequest,
      RawPaddlePostSubscriptionUsersResponse
    >(PADDLE_SUBSCRIPTION_USERS.url, PADDLE_SUBSCRIPTION_USERS.method, data).then((subscriptions) =>
      subscriptions.map(formatSubscription)
    )
  }

  async updateSubscription(
    data: PaddleSdkUpdateSubscriptionRequest<TPassthrough>
  ): Promise<PaddleSdkUpdateSubscriptionResponse> {
    return this.apiRequest<
      RawPaddlePostSubscriptionUsersUpdateRequest,
      RawPaddlePostSubscriptionUsersUpdateResponse
    >(PADDLE_SUBSCRIPTION_USERS_UPDATE.url, PADDLE_SUBSCRIPTION_USERS_UPDATE.method, {
      ...data,
      passthrough: this.stringifyPassthrough(data.passthrough),
    })
  }

  async cancelSubscription(
    data: PaddleSdkCancelSubscriptionRequest
  ): Promise<PaddleSdkCancelSubscriptionResponse> {
    return this.apiRequest<
      RawPaddlePostSubscriptionUsersCancelRequest,
      RawPaddlePostSubscriptionUsersCancelResponse
    >(PADDLE_SUBSCRIPTION_USERS_CANCEL.url, PADDLE_SUBSCRIPTION_USERS_CANCEL.method, data)
  }

  async createSubscriptionModifier(
    data: PaddleSdkCreateSubscriptionModifierRequest
  ): Promise<PaddleSdkCreateSubscriptionModifierResponse> {
    return this.apiRequest<
      RawPaddlePostSubscriptionModifiersCreateRequest,
      RawPaddlePostSubscriptionModifiersCreateResponse
    >(PADDLE_SUBSCRIPTION_MODIFIERS_CREATE.url, PADDLE_SUBSCRIPTION_MODIFIERS_CREATE.method, data)
  }
}
