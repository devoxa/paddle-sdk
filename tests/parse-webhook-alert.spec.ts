import { PUBLIC_KEY, WEBHOOK_BODY_SUBSCRIPTION_CREATED } from './fixtures'
import { PaddleSdk } from '../src/index'

function parse(publicKey: string, body: any) {
  const paddleSdk = new PaddleSdk({ publicKey, vendorId: 123, vendorAuthCode: 'FooBarBaz' })
  return paddleSdk.parseWebhookAlert(body)
}

describe('webhooks -> parse webhook alert', () => {
  it('parses a webhook correctly', () => {
    expect(parse(PUBLIC_KEY, WEBHOOK_BODY_SUBSCRIPTION_CREATED)).toMatchSnapshot()
  })
})
