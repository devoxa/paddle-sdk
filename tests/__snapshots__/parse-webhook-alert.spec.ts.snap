// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`webhooks -> parse webhook alert parses a webhook correctly 1`] = `
Object {
  "alert_id": 2091557455,
  "alert_name": "SUBSCRIPTION_CREATED",
  "cancel_url": "https://checkout.paddle.com/subscription/cancel?user=6&subscription=1&hash=acb15fc760b9d4cb4fae9abf67b7b4d52f12441e",
  "checkout_id": "2-461ecbe710cab0e-f2e284ef00",
  "currency": "GBP",
  "email": "gordon35@example.org",
  "event_time": 2020-08-09T22:11:09.000Z,
  "marketing_consent": true,
  "next_bill_date": 2020-09-03T00:00:00.000Z,
  "passthrough": Object {
    "foo": "bar",
  },
  "quantity": 23,
  "source": "https://genesis.devoxa.io/",
  "status": "TRIALING",
  "subscription_id": 6,
  "subscription_plan_id": 7,
  "unit_price": 9.99,
  "update_url": "https://checkout.paddle.com/subscription/update?user=3&subscription=3&hash=ce39e713407ccffbc867dba87e39e88324114274",
  "user_id": 9,
}
`;
